### PR TITLE
Gleam-from-Elm cheatsheet

### PR DESCRIPTION
**Pull request by michaeljones**
Starting a PR in case there is any early thoughts or feedback. No rush. I'll try to work through the rest this week.

---

Starting from the Elixir cheatsheet at Louis' recommendation. I've added
an overview to give a broad comparison before proceeding to edit the
document line by line to shape to Elm instead.

It is not completed. I have removed the unchanged parts for the moment
though the entries in the table of contents still correspond to them.

### CODE REVIEW

**Comment by michaeljones** [comment]
Just pushed another commit that sees the rest of the file converted.

I guess it is less than ideal as it is still goes feature by feature but through Elixir features and then explains the relevance to Elm & Gleam (without mentioned Elixir but that is the skeleton I've built off.) It presumably makes sense to restructure based on Elm features and copy the sections in where desirable. I'll continue to work on it.

**Comment by michaeljones** [comment]
Reorganised with a table of contents which is more Elm targeted. Plenty of TODOs though so much more work to be done. Looks more convincingly like an Elm dev's guide now though.

File: cheatsheets/gleam-for-elm-users.md [hunk 1]
```diff
@@ -0,0 +1,699 @@
...
+- [Package management](#package-management)
+- [Implementation](#implementation)
+- [Other concepts](#other-concepts)
+  - [Atoms](#atoms)
+
+## Overview
+
+Elm and Gleam have similar goals of providing a robust and sound type system with an approachable
+set of features. Gleam started off with a similar syntax to Elm but as moved to use one that is more
+familiar to many developers.
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:45]:
```suggestion
set of features.
```

An interesting bit of history, but not something I think belongs here. I wouldn't want any newcomer Elm syntax fans to feel let down by this change.

File: cheatsheets/gleam-for-elm-users.md [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+Comments starting with `///` are used to document the following statement. Comments starting with `////` are used to document the current module.
+
+```rust
+//// This module is very important.
+
+/// The answer to life, the universe, and everything.
+const answer: Int = 42
+```
+
+There are no multiline commends in Gleam.
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:95]:
```suggestion
There are no multiline comments in Gleam.
```

File: cheatsheets/gleam-for-elm-users.md [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+
+#### Gleam
+
+In Gleam comments are written with a `//` prefix.
+
+```rust
+// Hello, Joe!
+```
+
+Comments starting with `///` are used to document the following statement. Comments starting with `////` are used to document the current module.
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:86]:
```suggestion
Comments starting with `///` are used to document the following function, constant, or type definition. Comments starting with `////` are used to document the current module.
```

There was some confusion around the word "statement" in discord the other day. I am hoping this language is clearer.

File: cheatsheets/gleam-for-elm-users.md [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+let
+    size = 50
+    size = size + 100 -- Compile Error!
+in
+```
+
+#### Gleam
+
+Gleam has the `let` keyword before its variable names. You can re-assign variables and you can
+shadow variables from other scopes.
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:115]:
```suggestion
shadow variables from other scopes. This does not mutate the previously assigned value.
```

File: cheatsheets/gleam-for-elm-users.md [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+#### Elm
+
+In Elm, functions can **optionally** have their argument and return types annotated. These type annotations will always be checked by the compiler and throw a compilation error if not valid. The compiler will still type check your program using type inference if annotations are omitted.
+
+
+```elm
+sum : number -> number -> number
+sum x y = x + y
+
+@spec mul(number, number) :: boolean # no Elixir compile error
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:227]:
```suggestion
```

File: cheatsheets/gleam-for-elm-users.md [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+
+### Referencing functions
+
+Both Elm and Gleam have a single namespace for value and functions within a module, so there
+is no need for a special syntax to assign a module function to a variable.
+
+### Calling anonymous functions
+
+Neither Elm nor Gleam require special syntax for calling anonymous functions.
+
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:265]: Maybe we can remove this bit? Not sure Elm programmers would be expecting these features. What do you think?

```suggestion
```

File: cheatsheets/gleam-for-elm-users.md [hunk 7]
```diff
@@ -0,0 +1,699 @@
...
+In Elm, tuplies are limited to only 2 or 3 entries. It is recommended to use records for more larger numbers of entries.
+
+```elm
+myTuple = ("username", "password", 10)
+(_, password, _) = myTuple
+```
+
+#### Gleam
+
+There is no limit to the number of entries in Gleam tuples.
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:463]:
```suggestion
There is no limit to the number of entries in Gleam tuples, but records are still recommended as giving names to fields adds clarity.
```

File: cheatsheets/gleam-for-elm-users.md [hunk 8]
```diff
@@ -0,0 +1,699 @@
...
+Gleam's custom types can be used in much the same way. At runtime, they have a tuple representation and are compatible with Erlang records.
+
+```rust
+type Person {
+  Person(name: String, age: Int)
+}
+
+let person = Person(name: "Jake", age: 35)
+let name = person.name
+```
```
Comment (lpil) [inline cheatsheets/gleam-for-elm-users.md:550]: Not sure what is happening here, GitHub's syntax highlighting seems to have broken?

**Review by lpil — COMMENTED** [review]
This is fantastic work, thank you. Your writing is very clear. I've left some small notes but otherwise I'd love to merge this in when you're happy.

File: cheatsheets/gleam-for-elm-users.md (hunk 6 above)
Comment (michaeljones) [reply cheatsheets/gleam-for-elm-users.md:265]: Yeah, makes total sense. This a hang over from the original Elixir layout and it seemed odd to delete valid information but I can see it is potentially confusing if someone knows little about these things.

**Comment by michaeljones** [comment]
I've added your changes, completed the outstanding TODO sections and removed line breaks from paragraphs for consistency and to reduce diff noise on future edits. Though personally, I've not figured out how to get vim to wrap the lines at a reason width so I find this less readable :) 

I'm not the best at proof-reading my own work so edits are welcome.

I thought we might add a section on `todo` at some point as `Debug.todo` and `todo` exist as parallels. Not aiming to include that here myself.

I have attempted to link to the language tour in one place (for opaque types.) I'm not sure if the syntax of the link will work properly. 

Otherwise I'm happy with it to be merged if you're happy.

**Comment by lpil** [comment]
> I thought we might add a section on todo at some point as Debug.todo and todo exist as parallels. Not aiming to include that here myself.

Ah yes! That would be a good addition. Thank you

**Pull request merged**